db/db_impl/db_impl.cc

```diff
@@ -1688,35 +1695,49 @@ struct IterState {
   bool background_purge;
 };
 
-static void CleanupIteratorState(void* arg1, void* /*arg2*/) {
-  IterState* state = reinterpret_cast<IterState*>(arg1);
+static void CleanupSuperVersionHandle(void* arg1, void* /*arg2*/) {
+  SuperVersionHandle* sv_handle = reinterpret_cast<SuperVersionHandle*>(arg1);
 
-  if (state->super_version->Unref()) {
+  if (sv_handle->super_version->Unref()) {
     // Job id == 0 means that this is not our background process, but rather
     // user thread
     JobContext job_context(0);
 
-    state->mu->Lock();
-    state->super_version->Cleanup();
-    state->db->FindObsoleteFiles(&job_context, false, true);
-    if (state->background_purge) {
-      state->db->ScheduleBgLogWriterClose(&job_context);
-      state->db->AddSuperVersionsToFreeQueue(state->super_version);
-      state->db->SchedulePurge();
+    sv_handle->mu->Lock();
+    sv_handle->super_version->Cleanup();
+    sv_handle->db->FindObsoleteFiles(&job_context, false, true);
```

Comment:
Just wondering, why are `FindObsoleteFiles`, `ScheduleBgLogWriterClose` and `PurgeObsoleteFiles` done only in some last SuperVersion->Unref() cases but not all? For example, they are not done [here](https://github.com/facebook/rocksdb/blob/d23752f672abe0287c42cc2985dac43a61172845/db/column_family.cc#L1248) and [here](https://github.com/facebook/rocksdb/blob/d23752f672abe0287c42cc2985dac43a61172845/db/db_impl/db_impl.cc#L3759). Otherwise we could have a consistent way of doing the last Unref() of all SuperVersions.

Reply:
Not sure. It looks like those other approaches might not delete obsolete files or even trigger background deletion. I am not sure we've revisited that since background purging feature was introduced. Without it I guess file deletion on the read path would be surprising.

Reply:
The iterator cleanup involving file deletion came from here: a0ce3fd00a1ae35bcbb9edc603b7398d83a8eb23. The SuperVersion cleanup not involving file deletion came later: 539dd207df0283bbb218806cd317cd883ef21fff. There is a note that SST file deletion is missing but no reason why.

Reply:
Thanks for digging in to this. I guess for https://github.com/facebook/rocksdb/commit/539dd207df0283bbb218806cd317cd883ef21fff, one possible reason could be that since it's in `GetThreadLocalSuperVersion`, it assumes that purging could be done when this newer version of SuperVersion is destroyed?